README.md
# Tagesschau Downloader

Python script for playing and downloading all media on tagesschau.de with a single click.


I always imagined how nice it would be, if I got up and only needed a single click to enjoy my news during breakfast. And finally I came around to do it.

## Installation

Example for Ubuntu:
```bash
git clone https://github.com/hvoecking/tagesschau-downloader.git
cd tagesschau-downloader
# install pip3 if not already present, eg. sudo apt-get install python3-pip
sudo pip3 install virtualenv
virtualenv venv
source venv/bin/activate
python3 tagesschau-downloader.py [--help]
```

## Download

Using the standard options the ammount downloaded per day will average at about 80MB to 90MB.
Ranging from 20MB up to 150MB or more.

## License

MIT

## TODO

- [x] Fix formatting of last row on downloading
- [x] Do not touch files anymore
- [x] Do not split up in subfolders by date anymore to avoid downloading same files multiple times
- [x] Set track name in playlist as website title
- [x] Download to same directory per topic (cut off numbers at the end)
- [x] Command line options / settings file
  - [x] Help
  - [x] Audio/Video quality
  - [x] Directory for download (default: Save in downloads folder)
  - [x] Media Player
  - [x] Download manager
  - [x] Max age of files
- [x] Sort playlist by general article topic > time
- [x] Format code
- [x] Consistently sort video stuff before audio stuff


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.1.0 (2021-05-14)


### Features

* adapt to new url format ([be2d5f1](https://github.com/hvoecking/tagesschau-downloader/commit/be2d5f1d6c8d847ea381e43d4e71fc588e959bb8))
* adapt to new webpage format ([fcae9c0](https://github.com/hvoecking/tagesschau-downloader/commit/fcae9c066747abdf02766899e4c86f393f357016))
* setup build structure ([2351bef](https://github.com/hvoecking/tagesschau-downloader/commit/2351befb5daee141381802f465a3b0921a6d5ce4))


### Bug Fixes

* fix audio file pattern ([2e5a4b7](https://github.com/hvoecking/tagesschau-downloader/commit/2e5a4b7fc2c5a76ebb405f39f2e61ffcdfd23519))
* remove things that will not get implemented ([17b5ddc](https://github.com/hvoecking/tagesschau-downloader/commit/17b5ddc114a6e2b5083bb3410abdc52d6cf6b45c))

## 1.1.0 (2021-05-14)


### Features

* adapt to new url format ([be2d5f1](https://github.com/hvoecking/tagesschau-downloader/commit/be2d5f1d6c8d847ea381e43d4e71fc588e959bb8))
* adapt to new webpage format ([fcae9c0](https://github.com/hvoecking/tagesschau-downloader/commit/fcae9c066747abdf02766899e4c86f393f357016))
* setup build structure ([2351bef](https://github.com/hvoecking/tagesschau-downloader/commit/2351befb5daee141381802f465a3b0921a6d5ce4))


### Bug Fixes

* fix audio file pattern ([2e5a4b7](https://github.com/hvoecking/tagesschau-downloader/commit/2e5a4b7fc2c5a76ebb405f39f2e61ffcdfd23519))
* remove things that will not get implemented ([17b5ddc](https://github.com/hvoecking/tagesschau-downloader/commit/17b5ddc114a6e2b5083bb3410abdc52d6cf6b45c))

## 1.1.0 (2021-05-14)


### Features

* adapt to new url format ([be2d5f1](https://github.com/hvoecking/tagesschau-downloader/commit/be2d5f1d6c8d847ea381e43d4e71fc588e959bb8))
* adapt to new webpage format ([fcae9c0](https://github.com/hvoecking/tagesschau-downloader/commit/fcae9c066747abdf02766899e4c86f393f357016))
* setup build structure ([2351bef](https://github.com/hvoecking/tagesschau-downloader/commit/2351befb5daee141381802f465a3b0921a6d5ce4))


### Bug Fixes

* fix audio file pattern ([2e5a4b7](https://github.com/hvoecking/tagesschau-downloader/commit/2e5a4b7fc2c5a76ebb405f39f2e61ffcdfd23519))
* remove things that will not get implemented ([17b5ddc](https://github.com/hvoecking/tagesschau-downloader/commit/17b5ddc114a6e2b5083bb3410abdc52d6cf6b45c))


media-scraper.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  media-scraper.py
#
#  Copyright 2013 Heye Vöcking <heye.voecking at gmail.com>
#
#  This program is distributed under the terms of a slightly modified version
#  of the BSD License, please see the LICENSE.md file.
#

import os
import re
import sys
import copy
import argparse
from datetime import date

import urllib.parse
import urllib.error
import urllib.request
from bs4 import BeautifulSoup

import const as c

# From http://stackoverflow.com/questions/287871/print-in-terminal-with-colors-using-python
class bcolors:
    BOLD_SEQ = '\033[1m'
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'

    def disable(self):
        self.HEADER = ''
        self.OKBLUE = ''
        self.OKGREEN = ''
        self.WARNING = ''
        self.FAIL = ''
        self.ENDC = ''


def parse_args():
	NYI = "\nNOT YET IMPLEMENTED"

	parser = argparse.ArgumentParser()

	parser.add_argument("-v, --video-quality",
						choices=c.VIDEO_QUALITIES.keys(),
						default="webm",
						dest='video',
					    help="There are %d different video qualities "
							 "available for download: %s" %
							 (len(c.VIDEO_QUALITIES),
							  str(c.VIDEO_QUALITIES)))

	parser.add_argument("-a, --audio-quality",
						choices=c.AUDIO_QUALITIES.keys(),
						default="mp3",
						dest='audio',
					    help="There are %d different audio qualities "
							 "available for download." %
							 len(c.AUDIO_QUALITIES))

	parser.add_argument("-d, --download-directory",
						dest='dir',
						default="~/Downloads",
						help="The directory to which the files should "
							 "be downloaded.")

	parser.add_argument("-p, --player",
						dest='player',
						default="vlc",
						help="The media player which will be called "
							 "to open the generated playlist.")

	parser.add_argument("-f, --file-age",
						dest='age',
						default=0,
						help="The maximum number of days since the "
							 "file has been published. Use 0 for "
							 "unlimited." + NYI)

	parser.add_argument("--append",
						choices=[True, False],
						dest='append',
						default=True,
						help="Whether the files downloaded in this "
						     "session should be appended to the "
						     "previous playlist." + NYI)

	parser.add_argument("--cache-age",
						dest='c_age',
						default=0,
						help="Files with an age greater than the "
							 "specified number of days will be "
							 "deleted. The default is 0, which means "
							 "no means there is no age limit." + NYI)

	parser.add_argument("--cache-size",
						dest='c_size',
						default=0,
						help="The cache size specified in megabytes. "
							 "If this size is exceeded files will be "
							 "removed in chronological order. The "
							 "default is 0, which means: thee is no "
							 "size limit." + NYI)

	parser.add_argument("--download-tool",
						choices={"curl","wget"},
						dest='tool',
						default="curl",
						help="The command line tool used to download "
							 "the media files. Command line output "
							 "is optimized for curl.")

	default_ua = ("Mozilla/5.0 (Windows; U; Windows NT 6.1; ru; "
				  "rv:1.9.2.3) Gecko/20100401 Firefox/4.0 "
				  "(.NET CLR 3.5.30729)")
	parser.add_argument("-u, --user-agent",
						dest='ua',
						default=default_ua,
						help="The user agent to be used.")

	args = parser.parse_args()

	args.dir += "/"

	return args


def setup():

	c.TEXT_COLORS = {\
		"   " : (bcolors.ENDC, bcolors.ENDC, bcolors.ENDC), \
		"** " : (bcolors.OKBLUE, bcolors.OKBLUE, bcolors.OKBLUE), \
		"**~" : (bcolors.OKBLUE, bcolors.OKBLUE, bcolors.ENDC), \
		"  ~" : (bcolors.FAIL, bcolors.FAIL, bcolors.FAIL), \
		" * " : (bcolors.ENDC, bcolors.OKBLUE, bcolors.ENDC), \
		" *~" : (bcolors.ENDC, bcolors.OKBLUE, bcolors.OKBLUE), \
		"*  " : (bcolors.OKBLUE, bcolors.ENDC, bcolors.ENDC), \
		"* ~" : (bcolors.OKBLUE, bcolors.ENDC, bcolors.OKBLUE) \
	}

	c.PREFIX_TYPE = { " " : r"%2d", "*" : r"*%d", "-" : r"-%d" }

	"""
	Style of video URLs:

	http://download.media.tagesschau.de/
		video/2013/0526/TV-20130526-2257-3401.podm.h264.mp4
	http://download.media.tagesschau.de/
		video/2013/0526/TV-20130526-2257-3401.webm.h264.mp4
	http://download.media.tagesschau.de/
		video/2013/0526/TV-20130526-2257-3401.webm.webm
	http://download.media.tagesschau.de/
		video/2013/0526/TV-20130526-2257-3401.webl.h264.mp4
	"""

	c.PV = 'TV'

	c.VIDEO_QUALITIES = {
		'mobil' : "podm.h264.mp4",
		'mittel' : "webm.h264.mp4",
		'webm' : "webm.webm",
		'hoch' : "webl.h264.mp4"}

	c.PLAYLIST_TYPE = "m3u"

	"""
	Style of audio URLs:

	http://media.tagesschau.de/audio/2013/0526/AU-20130526-2327-3101.mp3
	http://media.tagesschau.de/audio/2013/0526/AU-20130526-2327-3101.ogg
	"""

	c.PA = 'AU'

	c.AUDIO_QUALITIES = {
		'mp3' : "mp3",
		'ogg' : "ogg"}

	c.URL_PATTERN = re.compile(r"\d*\.html")


	args = parse_args()


	c.HEADERS = {'User-Agent': args.ua}


	c.VIDEO_FILE_PATTERN = re.compile(r"http://download\.media\." +
		"tagesschau\.de/video/\d{4}\/\d{4}/%s-\d{8}-\d{4}-\d{4}\.%s" %
		(c.PV, c.VIDEO_QUALITIES[args.video]))

	c.AUDIO_FILE_PATTERN = re.compile(r"http://media\.tagesschau\.de/"
		"audio/\d{4}\/\d{4}/%s-\d{8}-\d{4}-\d{4}\.%s" %
		(c.PA, c.AUDIO_QUALITIES[args.audio]))



	print("Video quality: ", bcolors.OKBLUE, args.video, bcolors.ENDC)
	print("Audio quality: ", bcolors.OKBLUE, args.audio, bcolors.ENDC)
	print("Downloading to:", bcolors.OKBLUE, args.dir, bcolors.ENDC)
	print()

	return args

# Create a directory if it doesn't exist
def mkdir(directory):
	if not os.path.exists(directory):
		os.makedirs(directory)

# Saves the html (as string) to a file in dir
def save_as(html, dir, name):
	f = open(dir + name, 'w')
	f.write(str(html))
	f.close()

# Opens a url and returns it's page as a BeautifulSoup object
def cook_soup(opener, url, dir=None, name=None, data=None):
	request = urllib.request.Request(url, data, c.HEADERS)
	response = opener.open(request)

	html = response.read()
	if dir is not None and name is not None:
		save_as(html, dir, name)

	return BeautifulSoup(html)

def create_counter():
	counter = {}
	counter[c.PV] = 0
	counter[c.PA] = 0
	return counter

def print_table_row(videos, audios, title,
					v_prefix=" ", a_prefix=" ",
					t_prefix=" "):

	v_color, a_color, t_color = c.TEXT_COLORS[v_prefix + a_prefix + t_prefix]

	v_str = str(c.PREFIX_TYPE[v_prefix]) % videos
	a_str = str(c.PREFIX_TYPE[a_prefix]) % audios
	string = ("%s  %s  %s|%s  %s   %s|%s %s%s%s" %
			  (v_color, v_str, bcolors.ENDC,
			   a_color, a_str, bcolors.ENDC,
			   t_color, t_prefix, title, bcolors.ENDC))
	print (string)

def main(location, age, player, tool, feed):
	today = str(date.today())
	mkdir(location)

	playlist_file = location + today + "." + c.PLAYLIST_TYPE
	playlist = open(playlist_file, 'w'):
	playlist.write("#EXTM3U\n")

	# Will hold all media urls with their attributes
	medias = {}

	count = create_counter()
	duplicates = 0

	# Prepare the opener
	opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(None))

	soup = cook_soup(opener, feed)

	urls = []
	for entry in soup.find_all('entry'):
		urls.append(entry.link['href'])

	print("Opening", len(urls), "articles on", feed, "|", today)
	print()
	print("%svideo %s| %saudio %s|%s Title%s" %
		  (bcolors.BOLD_SEQ, bcolors.ENDC,
		   bcolors.BOLD_SEQ, bcolors.ENDC,
		   bcolors.BOLD_SEQ, bcolors.ENDC))
	print("------|-------|---------------")

	# Now look on each page if video or audio files are present
	for url in urls:

		if not url.endswith(".html"):
			print ("Skipping non-html document, with url", url)
			continue
		if not url.startswith("http://www.tagesschau.de"):
			print ("Skipping non-tagesschau url", url)
			continue

		name = url.split("/")[-1]
		general_topic = re.sub(c.URL_PATTERN, "", name)
		dir = location + general_topic + "/"
		mkdir(dir)

		soup = cook_soup(opener, url, dir, name)

		videos = soup.find_all(attrs={'href': c.VIDEO_FILE_PATTERN})
		audios = soup.find_all(attrs={'href': c.AUDIO_FILE_PATTERN})

		title = soup.find('title').get_text().split(" | ")[0]

		# Insert the found media files into their corresponding arrays

		media_url = videos + audios

		before = copy.copy(count)

		for m in media_url:
			m_url = m['href']
			file = m_url.split("/")[-1]
			prefix = file[:2]
			key = file[3:]
			if not (key in medias or (os.path.exists(dir + file) and os.path.getsize(dir + file) > 0)):
				count[prefix] += 1
				time = ("%s.%s.%s %s:%s" %
					    (file[9:11], file[7:9], file[3:7], file[12:14],
					     file[14:16]))
				medias[key] = (dir, file, m_url,
							   "(%s) %s: %s" %
							   (prefix, time, title),
							   general_topic)
			else:
				duplicates += 1

		len_v = len(videos)
		count_v = count[c.PV] - before[c.PV]
		len_a = len(audios)
		count_a = count[c.PA] - before[c.PA]

		if len_v + len_a == 0:
			print_table_row(len_v, len_a, title, " ", " ", "~")
		else:
			pv = "*" if len_v != count_v else " "
			pa = "*" if len_a != count_a else " "
			pt = "~" if len_a * len_v != 0 and before == count else " "
			print_table_row(count_v, count_a, title, pv, pa, pt)

	# We will prepare a command (that will be executed via os.system)
	# in this string

	# At first we are going to start vlc with a delay of 1 second, so
	# the file download can already start in the background. The output
	# vlc is sent to /dev/null so it doesn't interfere with the wget
	# output
	cmd = ["sleep 1 && vlc \"%s\" > /dev/null 2>&1 > /dev/null & " %
		  playlist_file]

	print("------|-------|---------------")
	print_table_row(count[c.PV], count[c.PA], "will be downloaded...");
	if duplicates != 0:
		print("%s* %d files have been skipped because they were "
			  "duplicates...%s" %
			  (bcolors.OKBLUE, duplicates, bcolors.ENDC))


	# Sort the media keys to be grouped by topic
	sorted_medias = {}
	for key in medias.keys():
		general_topic = medias[key][4]
		sorted_medias[general_topic + key] = medias[key]

	keys = sorted(sorted_medias.keys())
	medias = sorted_medias

	# Add for all video and audio files that are not present in the
	# filesystem an entry in the plylist and a call to wget, so it will
	# be downloaded
	for key in keys:
		e = medias[key]
		dir = e[0]
		file = e[1]
		url = e[2]
		title = e[3]
		path = dir + file
		escaped_title = title.replace("-", "‑")
		playlist.write("#EXTINF:,,%s\n" % escaped_title)
		playlist.write(path + "\n")
		option = ""
		grep = ""
		if tool == 'curl':
			option = "-o"
			grep = "-v Total"
		elif tool == 'wget':
			option = "-O"
			grep =  "saved"
		cmd.append("cd %s && %s %s %s %s 2>&1 | grep %s && " %
				   (dir, tool, url, option, file, grep))

	playlist.close()


	command = ''.join(cmd) + "echo done"

	os.system(command)

	return

if __name__ == '__main__':
	print(bcolors.BOLD_SEQ + "Media Scraper for tagesschau.de" +
		bcolors.ENDC)
	print(" (c) by" + bcolors.BOLD_SEQ + " Heye Voecking " +
		bcolors.ENDC + bcolors.OKBLUE + \
		"<heye.voecking+mediascraper[at]gmail[dot]com>" + bcolors.ENDC)
	print()
	print(bcolors.FAIL + \
		"This program is distributed WITHOUT ANY WARRANTY!" +
		bcolors.ENDC)
	print()

	args = setup()

	main(args.dir, args.age, args.player, args.tool,
		 "http://www.tagesschau.de/xml/atom/")


tagesschau-downloader.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  Copyright 2013 Heye Vöcking <heye.voecking at gmail.com>
#
#  This program is distributed under the terms of a slightly modified version
#  of the BSD License, please see the LICENSE.md file.
#

import argparse
import copy
import json
import os
import re
import urllib.error
import urllib.parse
import urllib.request
from datetime import date
from html import unescape

from bs4 import BeautifulSoup

BOLD = "\033[1m"
HEADER = "\033[95m"
BLUE = "\033[94m"
GREEN = "\033[92m"
YELLOW = "\033[93m"
RED = "\033[91m"
END = "\033[0m"

TEXT_COLORS = {
    "   ": (END, END, END),
    "** ": (BLUE, BLUE, BLUE),
    "**~": (BLUE, BLUE, END),
    "  ~": (RED, RED, RED),
    " * ": (END, BLUE, END),
    " *~": (END, BLUE, BLUE),
    "*  ": (BLUE, END, END),
    "* ~": (BLUE, END, BLUE),
}

PREFIX_TYPE = {" ": r"%2d", "*": r"*%d", "-": r"-%d"}

PV = "TV"
PA = "AU"

URL_PATTERN = re.compile(r"\d*\.html")
AUDIO_QUALITIES = {"mp3": "mp3", "ogg": "ogg"}
VIDEO_QUALITIES = {
    "sm": "websm.h264.mp4",
    "m": "webm.h264.mp4",
    "l": "webl.h264.mp4",
    "xl": "webxl.h264.mp4",
}

PLAYLIST_TYPE = "m3u"


def compile_pattern(type, prefix, quality):
    return re.compile(
        r"//download\.media\.tagesschau\.de/%s/\d{4}\/\d{4}/%s-\d{8}-\d{4}-\d{4}\.%s"
        % (type, prefix, quality)
    )


def parse_args():
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "-v, --video-quality",
        choices=VIDEO_QUALITIES.keys(),
        default="l",
        dest="video",
        help=" ".join(
            [
                f"There are {len(VIDEO_QUALITIES)} different video qualities available for",
                f"download: {str(VIDEO_QUALITIES)}",
            ]
        ),
    )

    parser.add_argument(
        "-a, --audio-quality",
        choices=AUDIO_QUALITIES.keys(),
        default="mp3",
        dest="audio",
        help=" ".join(
            [
                f"There are {len(AUDIO_QUALITIES)} different audio qualities available for",
                "download.",
            ]
        ),
    )

    parser.add_argument(
        "-d, --download-directory",
        dest="dir",
        default="~/Downloads",
        help="The directory to which the files should be downloaded.",
    )

    parser.add_argument(
        "-p, --player",
        dest="player",
        default="vlc",
        help="The media player which will be called to open the generated playlist.",
    )

    parser.add_argument(
        "--download-tool",
        choices={"curl", "wget"},
        dest="tool",
        default="curl",
        help=" ".join(
            [
                "The command line tool used to download the media files. Command line",
                "output is optimized for curl.",
            ]
        ),
    )

    default_ua = " ".join(
        [
            "Mozilla/5.0 (Windows; U; Windows NT 6.1; ru; rv:1.9.2.3)",
            "Gecko/20100401",
            "Firefox/4.0 (.NET CLR 3.5.30729)",
        ]
    )
    parser.add_argument(
        "-u, --user-agent",
        dest="ua",
        default=default_ua,
        help="The user agent to be used.",
    )

    return parser.parse_args()


# Create a directory if it doesn't exist
def mkdir(directory):
    if not os.path.exists(directory):
        os.makedirs(directory)


# Opens a url and returns it's page as a BeautifulSoup object
def cook_soup(opener, url, headers, dir=None, name=None, data=None):
    request = urllib.request.Request(url, data, headers)
    response = opener.open(request)

    resp = response.read().decode("utf-8")
    html = unescape(resp)
    if dir is not None and name is not None:
        with open(dir + name, "w") as f:
            f.write(str(html))

    return BeautifulSoup(html, "html.parser")


def print_table_row(videos, audios, title, v_prefix=" ", a_prefix=" ", t_prefix=" "):
    v_color, a_color, t_color = TEXT_COLORS[v_prefix + a_prefix + t_prefix]

    v_str = str(PREFIX_TYPE[v_prefix]) % videos
    a_str = str(PREFIX_TYPE[a_prefix]) % audios
    print(
        "|".join(
            [
                f"{v_color}  {v_str}  {END}",
                f"{a_color}  {a_str}   {END}",
                f"{t_color} {t_prefix}{title}{END}",
            ]
        )
    )


def main(location, tool, feed, headers, VIDEO_FILE_PATTERN, AUDIO_FILE_PATTERN):
    today = str(date.today())
    mkdir(location)

    playlist_file = location + today + "." + PLAYLIST_TYPE
    playlist = open(playlist_file, "w")
    playlist.write("#EXTM3U\n")

    # Will hold all media urls with their attributes
    medias = {}

    count = {
        PV: 0,
        PA: 0,
    }
    duplicates = 0

    # Prepare the opener
    opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(None))

    soup = cook_soup(opener, feed, headers)

    urls = []
    for entry in soup.find_all("entry"):
        urls.append(entry.link["href"])

    print(f"Opening {len(urls)} articles on {feed} | {today}")
    print()
    print(f"{BOLD}video {END}| {BOLD}audio {END}|{BOLD} Title{END}")
    print("------|-------|---------------")

    # Now look on each page if video or audio files are present
    for url in urls:

        if not url.endswith(".html"):
            print(f"Skipping non-html document, with url {url}")
            continue
        if not url.startswith("https://www.tagesschau.de"):
            print(f"Skipping non-tagesschau url {url}")
            continue

        name = url.split("/")[-1]
        general_topic = re.sub(URL_PATTERN, "", name)
        dir = f"{location}{general_topic}/"
        mkdir(dir)
        try:
            soup = cook_soup(opener, url, headers, dir, name)
        except urllib.error.URLError as err:
            print(f"Failed to open '{url}':{err}")
            continue

        videos = soup.find_all(attrs={"data-config": VIDEO_FILE_PATTERN})
        audios = soup.find_all(attrs={"data-config": AUDIO_FILE_PATTERN})

        title = soup.find("title").get_text().split(" | ")[0]

        # Insert the found media files into their corresponding arrays
        media_url = videos + audios

        before = copy.copy(count)
        for m in media_url:
            m_url = json.loads(m["data-config"])["mc"]["_mediaArray"][0][
                "_mediaStreamArray"
            ][-1]["_stream"]
            if not m_url.startswith("http"):
                m_url = "https:" + m_url
            file = m_url.split("/")[-1]
            prefix = file[:2]
            key = file[3:]
            path = dir + file
            if key in medias or (os.path.exists(path) and os.path.getsize(path) > 0):
                duplicates += 1
                continue

            count[prefix] += 1
            time = f"{file[9:11]}.{file[7:9]}.{file[3:7]} {file[12:14]}:{file[14:16]}"
            medias[key] = (
                dir,
                file,
                m_url,
                f"({prefix}) {time}: {title}",
                general_topic,
            )

        len_v = len(videos)
        count_v = count[PV] - before[PV]
        len_a = len(audios)
        count_a = count[PA] - before[PA]

        if len_v + len_a == 0:
            print_table_row(len_v, len_a, title, " ", " ", "~")
        else:
            pv = "*" if len_v != count_v else " "
            pa = "*" if len_a != count_a else " "
            pt = "~" if len_a * len_v != 0 and before == count else " "
            print_table_row(count_v, count_a, title, pv, pa, pt)

    # We will prepare a command (that will be executed via os.system)
    # in this string

    # At first we are going to start vlc with a delay of 1 second, so
    # the file download can already start in the background. The output
    # vlc is sent to /dev/null so it doesn't interfere with the wget
    # output
    cmd = [f'sleep 1 && vlc "{playlist_file}" > /dev/null 2>&1 > /dev/null & ']

    print("------|-------|---------------")
    print_table_row(count[PV], count[PA], "will be downloaded...")
    if duplicates != 0:
        print(f"{BLUE}* {duplicates} duplicates skipped...{END}")

    # Sort the media keys to be grouped by topic
    sorted_medias = {}
    for key in medias.keys():
        general_topic = medias[key][4]
        sorted_medias[general_topic + key] = medias[key]

    keys = sorted(sorted_medias.keys())
    medias = sorted_medias

    # Add for all video and audio files that are not present in the
    # filesystem an entry in the plylist and a call to wget, so it will
    # be downloaded
    for key in keys:
        e = medias[key]
        dir = e[0]
        file = e[1]
        url = e[2]
        title = e[3]
        path = dir + file
        escaped_title = title.replace("-", "‑")
        playlist.write(f"#EXTINF:,,{escaped_title}\n")
        playlist.write(path + "\n")
        option = ""
        grep = ""
        if tool == "curl":
            option = "-o"
            grep = "-v Total"
        elif tool == "wget":
            option = "-O"
            grep = "saved"
        cmd.append(f"cd {dir} && {tool} {url} {option} {file} 2>&1 | grep {grep} && ")

    playlist.close()
    os.system("".join(cmd) + "echo done")


if __name__ == "__main__":
    args = parse_args()

    print(f"{BOLD}Downloader for tagesschau.de{END}")
    print(
        f" (c) by{BOLD} Heye Voecking {END}{BLUE}<heye.voecking[at]gmail[dot]com>{END}"
    )
    print()
    print(f"{RED}This program is distributed WITHOUT ANY WARRANTY!{END}")
    print()
    print(f"Video quality:  {BLUE}{args.video}{END}")
    print(f"Audio quality:  {BLUE}{args.audio}{END}")
    print(f"Downloading to: {BLUE}{args.dir}{END}")
    print()

    main(
        args.dir + "/",
        args.tool,
        "http://www.tagesschau.de/xml/atom/",
        {"User-Agent": args.ua},
        compile_pattern("video", PV, VIDEO_QUALITIES[args.video]),
        compile_pattern("audio", PA, AUDIO_QUALITIES[args.audio]),
    )
